common/src/address.ts

```diff
@@ -0,0 +1,173 @@
+  [/ center /g, " "],
+  [/ pl /g, " "],
+  [/ place /g, " "],
+  [/ plz /g, " "],
+  [/ plaza /g, " "],
+  [/ pkw?y /g, " "],
+  [/ parkway /g, " "],
+  [/ cswy /g, " "],
+  [/ causeway /g, " "],
+  [/ byp /g, " "],
+  [/ bypass /g, " "],
+  [/ mall /g, " "],
+  [/ (xing|crssng) /g, " "],
+  [/ crossing /g, " "],
+  [/ sq /g, " "],
+  [/ square /g, " "],
+  [/ trl? /g, " "],
+  [/ trail /g, " "],
+  [/ (twp|twsp|townsh(ip)?) /g, " "],
+  [/ est(ate)? /g, " estates "],
+  [/ vlg /g, " "],
+  [/ village /g, " "],
+  [/ (ste|suite|unit|apt|apartment) #?(\d+) /g, " $2 "],
+  [/ (bld|bldg) #?(\d+) /g, " $2 "],
+  [/ #?(\d+) /g, " $1 "],
+  [/ (&|and) /g, " "],
+  // "First" - "Tenth" are pretty common (this could obviously go farther).
+  [/ first /g, " 1st "],
+  [/ second /g, " 2nd "],
+  [/ third /g, " 3rd "],
+  [/ fourth /g, " 4th "],
+  [/ fifth /g, " 5th "],
+  [/ sixth /g, " 6th "],
+  [/ seventh /g, " 7th "],
+  [/ eighth /g, " 8th "],
+  [/ ninth /g, " 9th "],
+  [/ tenth /g, " 10th "],
+];
+
+/**
+ * Simplify a text string (especially an address) as much as possible so that
+ * it might match with a similar string from another source.
+ */
+export function matchable(text: string): string {
+  return text
+    .toLowerCase()
+    .replace(POSSESSIVE_PATTERN, " ")
+    .replace(PUNCTUATION_PATTERN, " ")
+    .replace(MULTIPLE_SPACE_PATTERN, " ")
+    .trim();
+}
+
+export function matchableAddress(
+  text: string | string[],
+  line: number = null
+): string {
+  let lines = Array.isArray(text)
+    ? text
+    : text.split(ADDRESS_LINE_DELIMITER_PATTERN);
+
+  // If there are multiple lines and it looks like the first line is the name
+  // of a place (rather than the street), drop the first line.
+  if (lines.length > 1 && !/\d/.test(lines[0])) {
+    lines = lines.slice(1);
+  }
+
+  if (line != null) {
+    lines = lines.slice(line, line + 1);
+  }
+
+  let result = matchable(lines.join(" "));
+  for (const [pattern, expansion] of ADDRESS_EXPANSIONS) {
+    result = result.replace(pattern, expansion);
+  }
+
+  return result.replace(MULTIPLE_SPACE_PATTERN, " ").trim();
+}
+
+export interface UsAddress {
+  lines: string[];
+  city: string;
+  state: string;
+  zip: string | undefined;
+}
+
+/**
+ * Parse a US-style address string.
+ */
+export function parseUsAddress(address: string): UsAddress {
+  const match = address.match(ADDRESS_PATTERN);
+
+  // Detect whether we have something formatted like an address, but with
+  // obviously incorrect street/city/zip data, e.g. "., ., CA 90210".
+  const invalidMatch =
+    !match ||
+    match[1].replace(PUNCTUATION_PATTERN, "") === "" ||
+    match[2].replace(PUNCTUATION_PATTERN, "") === "" ||
+    match[4].replace(PUNCTUATION_PATTERN, "") === "";
+  if (invalidMatch) {
+    throw new ParseError(`Could not parse address: "${address}"`);
+  }
+
+  const zip = match[4];
+  if (zip.split("-")[0].length < 5) {
+    throw new ParseError(`Invalid ZIP code in address: "${address}"`);
```

Comment:
Minor note: this used to log a warning and set `zip` to `undefined` instead of throwing. However, the common module doesn’t have access to the same Sentry logging tools as the loader, so there wasn’t an easy way to do that (arguably there should be some common logging infrastructure in the univaf-common module, but that’s a whole other ball of wax).

Gracefully handling invalid zip codes was mainly an issue with The old MHealth Coach system that Albertsons used, and which they don’t use anymore, so this shouldn’t pose an actual issue in practice.